artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java

```diff
@@ -1971,7 +1988,9 @@ private synchronized int iterQueue(final int flushLimit,
 
             if (filter1 == null || filter1.match(ref.getMessage())) {
                messageAction.actMessage(tx, ref);
-               iter.remove();
+               if (remove) {
```

Comment:
Do we need the same check for scheduledDelivery messages (https://github.com/apache/activemq-artemis/blob/master/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ScheduledDeliveryHandlerImpl.java#L128) and paged messages (https://github.com/apache/activemq-artemis/blob/master/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java#L2002)?